File: engine/src/main/java/de/hanno/hpengine/engine/scene/VertexIndexBuffer.java
package de.hanno.hpengine.engine.scene;

import de.hanno.hpengine.engine.graphics.GpuContext;
import de.hanno.hpengine.engine.model.DataChannels;
import de.hanno.hpengine.engine.model.IndexBuffer;
import de.hanno.hpengine.engine.model.VertexBuffer;
import org.lwjgl.BufferUtils;

import java.io.Serializable;
import java.util.EnumSet;

public class VertexIndexBuffer implements Serializable {
    private volatile VertexBuffer vertexBuffer;
    private volatile IndexBuffer indexBuffer;
    private int currentBaseVertex = 0;
    private int currentIndexOffset = 0;

    public VertexIndexBuffer(GpuContext gpuContext, int vertexBufferSizeInFloatsCount, int indexBufferSizeInIntsCount, EnumSet<DataChannels> channels) {
        vertexBuffer = new VertexBuffer(gpuContext, BufferUtils.createFloatBuffer(vertexBufferSizeInFloatsCount), channels);
        indexBuffer  = new IndexBuffer(gpuContext, BufferUtils.createIntBuffer(indexBufferSizeInIntsCount));
    }

    public VertexBuffer getVertexBuffer() {
        return vertexBuffer;
    }

    public void setVertexBuffer(VertexBuffer vertexBuffer) {
        this.vertexBuffer = vertexBuffer;
    }

    public IndexBuffer getIndexBuffer() {
        return indexBuffer;
    }

    public void setIndexBuffer(IndexBuffer indexBuffer) {
        this.indexBuffer = indexBuffer;
    }

    public synchronized VertexIndexOffsets allocate(int elementsCount, int indicesCount) {
        VertexIndexOffsets vertexIndexOffsets = new VertexIndexOffsets(currentBaseVertex, currentIndexOffset);
        currentBaseVertex += elementsCount;
        currentIndexOffset += indicesCount;
        return vertexIndexOffsets;
    }

    public void resetAllocations() {
        currentBaseVertex = 0;
        currentIndexOffset = 0;
    }

    public static class VertexIndexOffsets {
        public final int vertexOffset;
        public final int indexOffset;

        public VertexIndexOffsets(int vertexOffset, int indexOffset) {
            this.vertexOffset = vertexOffset;
            this.indexOffset = indexOffset;
        }

    }
}

File: engine/src/main/java/de/hanno/hpengine/engine/scene/VertexIndexBuffer.kt
package de.hanno.hpengine.engine.scene

import AnimatedVertexStruktPackedImpl.Companion.type
import VertexStruktPackedImpl.Companion.type
import de.hanno.hpengine.engine.graphics.GpuContext
import de.hanno.hpengine.engine.graphics.renderer.pipelines.PersistentMappedBuffer
import de.hanno.hpengine.engine.graphics.renderer.pipelines.PersistentMappedStructBuffer
import de.hanno.hpengine.engine.graphics.renderer.pipelines.typed
import de.hanno.hpengine.engine.vertexbuffer.IndexBuffer
import org.lwjgl.BufferUtils

class VertexIndexBuffer(gpuContext: GpuContext<*>,
                        indexBufferSizeInIntsCount: Int) {

    var indexBuffer = IndexBuffer(gpuContext, BufferUtils.createIntBuffer(indexBufferSizeInIntsCount))
    private var currentBaseVertex = 0
    private var currentIndexOffset = 0

    fun allocate(elementsCount: Int, indicesCount: Int): VertexIndexOffsets = synchronized(this) {
        VertexIndexOffsets(currentBaseVertex, currentIndexOffset).apply {
            currentBaseVertex += elementsCount
            currentIndexOffset += indicesCount
        }
    }

    fun resetAllocations() {
        currentBaseVertex = 0
        currentIndexOffset = 0
    }

    var vertexStructArray = PersistentMappedBuffer(VertexStruktPacked.type.sizeInBytes, gpuContext).typed(VertexStruktPacked.type)
    var animatedVertexStructArray = PersistentMappedBuffer(AnimatedVertexStruktPacked.type.sizeInBytes, gpuContext).typed(AnimatedVertexStruktPacked.type)

    data class VertexIndexOffsets(val vertexOffset: Int, val indexOffset: Int)
}